Pl 130215 k3s roles

dpausp opened the pull request:
@flyingcircusio/release-managers

## Release process

Impact:

Changelog:

* OpenVPN: client example config is now accessible for users with `login` permissions.
* Kubernetes: replace old implementation based on NixOS k8s support with a k3s-based solution which is easier to manage and has some improvements like proper support for services of type LoadBalancer, a local path provisioner and pre-installed helm for package management, a metrics-server, and traefik for ingress. The roles k3s-server and k3s-agent replace kubernetes-master and kubernetes-node, respectively. The kubernetes-frontend role was removed, webgateway now automatically activates Kubernetes support when it's running in a project with a k3s-server. The k3s roles are currently in beta state. Documentation has been updated to reflect the changes and give more frontend configuration examples (#PL-130215).

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - access control API: not reachable from public network, permissions granted by client certificate
  - access control dashboard: public, protected by HTTPS, login/permissions granted by kubeconfig/token
  - config DB Postgres only reachable for k3s server
  - service/pod networks only internally accessible 
  - worker nodes work without public fe addresses, public traffic is handled by a frontend VM that talks to the internal  service/pod networks
  - other than that, we rely on k3s defaults and further security implications depend on Kubernetes configuration
- [x] Security requirements tested? (EVIDENCE)
  - checked manually in our k3s test cluster: dashboard API can only accessed with authentication locally and via nginx; API is listening to 6443 on all interfaces but firewalled from the outside; Postgres uses our default service config; config DB only accessible by root, k3s server runs with root permissions
  - automated test is still TODO

dpausp committed:
k3s-agent, -server roles, k3s-frontend service

 #PL-130215

dpausp committed:
openvpn: allow login group to read the openvpn client config

ctheune committed:
kubernetes/k3s: rework documentation

 #PL-130215

ctheune commented on nixos/roles/external_net/openvpn.nix:
```diff
@@ -288,6 +288,8 @@ in
 
         # generate client config (depends on results from pki.generate)
         cat > ${clientConfigFile} << EOF
+        chgrp login ${clientConfigFile}
```
changelog + public docs

ctheune commented on nixos/roles/k3s/agent.nix:
```diff
@@ -0,0 +1,94 @@
+
+      flyingcircus.activationScripts.kubernetes-apitoken-node = ''
+        mkdir -p /var/lib/k3s
+        umask 077
```
please wrap the snippet in a subshell `( )` OR make _our_ activationscripts get wrapped automatically. you could then also refactor the wireguard snippet

ctheune commented on nixos/roles/k3s/server.nix:
```diff
@@ -0,0 +1,240 @@
+    flyingcircus.activationScripts.k3s-apitoken =
+      lib.stringAfter [ "users" ] ''
+        mkdir -p /var/lib/k3s
+        umask 077
```
remember the `()` wrapper

ctheune commented on nixos/roles/k3s/server.nix:
```diff
@@ -0,0 +1,240 @@
+        enableACME = true;
+        serverAliases = tail addresses;
+        extraConfig = ''
+          auth_basic "FCIO";
```
vpn without basic auth + option flag to  completely prohibit access from outside

ctheune replied:
and use the .sudo-srv file

ctheune commented on nixos/services/k3s/frontend.nix:
```diff
@@ -0,0 +1,239 @@
+      default = {};
+      description = ''
+        Configure the frontend haproxy to forward traffic to cluster services/pods.
+
```
newline

ctheune commented on nixos/services/k3s/frontend.nix:
```diff
@@ -0,0 +1,239 @@
+  config = lib.mkIf cfg.enable {
+
+      flyingcircus.activationScripts.kubernetes-apitoken-node = ''
+        mkdir -p /var/lib/k3s
```
`()` wrapper

ctheune reviewed (changes requested):
Looks good, please adjust according to the comments.

dpausp committed:
Wrap fc activation scripts in subshells

Prevents leakage or variables and umask settings, for example.

dpausp committed:
openvpn: move accessNets from let to option

This is helpful if we want to reference the configured access nets
from Nix code.

dpausp committed:
k3s/dashboard: switch to token-based auth

We used basic auth before but it's very slow because the dashboard
makes a lot of requests and hashing in nginx takes ages with newer passwords.

 #PL-130215

dpausp committed:
Remove old kubernetes roles

 #PL-130215
